librbd: implement ordering for overlapping IOs

MahatiC opened the pull request:
..and block flushes until previous writes are completed.

Signed-off-by: Mahati Chamarthy <mahati.chamarthy@intel.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

MahatiC commented:
@dillaman ping. Could you please review this?

dillaman commented:
@MahatiC I have a branch building w/ this PR right now so that I can try to reproduce what you are seeing tomorrow.

dillaman commented:
@MahatiC I've run the test repeatedly for over an hour w/o hitting the failure you witnessed. Any chance you can provide debug logs from one of your failed runs?

MahatiC commented:
@dillaman The error occurs when this patch is applied on top of https://github.com/ceph/ceph/pull/24483 
Here's the error - https://paste.opensuse.org/view/raw/27332960

dillaman commented:
Thanks. After applying the other PR now it crashes immediately for me (and I see lockdep error due to swapped lock ordering):

> /home/jdillaman/ceph_wip/src/librbd/io/AioCompletion.cc: In function 'void librbd::io::AioCompletion::start_op()' thread 7f42273fe700 time 2019-07-17T09:28:48.808104-0400
> /home/jdillaman/ceph_wip/src/librbd/io/AioCompletion.cc: 143: FAILED ceph_assert(!async_op.started())
>  ceph version v15.0.0-2727-g504f25832c (504f25832cfe598e250f42f10aa595e0e308e60d) octopus (dev)
>  1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x14f) [0x7f423fe4bad1]
>  2: (()+0x273cbb) [0x7f423fe4bcbb]
>  3: (()+0x10d349d) [0x558d5541e49d]
>  4: (librbd::io::ImageRequestWQ<librbd::ImageCtx>::_void_dequeue()+0x618) [0x558d54e5ee88]
>  5: (ThreadPool::worker(ThreadPool::WorkThread*)+0x4db) [0x7f423ff4480b]
>  6: (ThreadPool::WorkThread::entry()+0x10) [0x7f423ff45d10]
>  7: (()+0x85a2) [0x7f42489f55a2]
>  8: (clone()+0x43) [0x7f423f3d9303]

I'll take a look through your log.

dillaman commented:
@MahatiC Can you provide the "debug rbd=20" logs prior to the crash and not just the crash backtrace?

MahatiC commented:
@dillaman Thanks. Never encountered the lockdep failure. And I had to run `TestLibRBD.FlushAio` multiple times to get the below logs as well. And I did see `ceph_assert(!async_op.started())` failure before but once I ensured locking when releasing blocked ios, it didn't occur. I'll look at that again.

I'm attaching a couple of debug logs of the backtrace I provided earlier. One log where librbd, rbd and objectcacher logs are set to 20 and another with just debug_rbd=20.
 
[librbd_rbd_objcacher.log](https://github.com/ceph/ceph/files/3404846/librbd_rbd_objcacher.log)
[rbd.log](https://github.com/ceph/ceph/files/3404848/rbd.log)

dillaman commented:
@MahatiC I don't see the crashes in those two logs. It's hard for me to track the offending IO if I can't see where to backtrace the logs from.

MahatiC commented:
> @MahatiC I don't see the crashes in those two logs. It's hard for me to track the offending IO if I can't see where to backtrace the logs from.

That's really all the logs are. The error is in the backtrace (that shows up on the CLI) and the logs just show that it crashed mid point while executing an IO (with no backtrace _within_ the logs. You should be able to reproduce this locally with "RBD_FEATURES=125 ./bin/unittest_librbd --gtest_filter=TestLibRBD=FlushAio --gtest_repeat=20". Running that a few times will throw this error.

dillaman commented:
> You should be able to reproduce this locally

That's the thing, the code is broken for me and immediately crashes on that `ceph_assert(!async_op.started())` failure. If you can fix that issue, I can run it myself (after disabling lockdep to avoid the other bug).

MahatiC commented:
@dillaman I merged them (#24483 & prev version of this PR) both. Hope this helps to reproduce. Let me know. Thanks!

dillaman commented:
@MahatiC I'm able to repeat the crash now (but I also still see the lockdep issue). I will investigate this tomorrow. Thx.

dillaman commented:
@MahatiC The cause of the crash is that `ImageRequestWQ::unblock_overlapping_io` method is incorrectly unblocking (now) overlapping IOs:

> 2019-07-26T10:06:30.117-0400 7f59bb940700 20 librbd::io::ImageRequestWQ: 0x55c1b38874c0 unblock_overlapping_io: unblocking off: 514118, len: 512
> 2019-07-26T10:06:30.117-0400 7f59bb940700 20 librbd::io::ImageRequestWQ: 0x55c1b38874c0 queue: ictx=0x55c1b38754c0, req=0x7f596c007f40
> 2019-07-26T10:06:30.125-0400 7f59bc141700 20 librbd::io::ImageRequestWQ: 0x55c1b38874c0 process: ictx=0x55c1b38754c0, req=0x7f596c007f40
> 2019-07-26T10:06:30.125-0400 7f59bc141700 20 librbd::io::ImageRequest: 0x7f59bc13d570 send: aio_write: ictx=0x55c1b38754c0, completion=0x55c1b38cb130
> 2019-07-26T10:06:30.125-0400 7f59bc141700 20 librbd::io::AioCompletion: 0x55c1b38cb130 set_request_count: pending=1
> 2019-07-26T10:06:30.125-0400 7f59c214d700 20 librbd::Journal: 0x7f59a4001cf0 append_io_events: event=AioWrite, offset=514118, length=512, flush=0, tid=157
> 
> 
> 2019-07-26T10:06:30.117-0400 7f59bb940700 20 librbd::io::ImageRequestWQ: 0x55c1b38874c0 unblock_overlapping_io: unblocking off: 514141, len: 512
> 2019-07-26T10:06:30.117-0400 7f59bb940700 20 librbd::io::ImageRequestWQ: 0x55c1b38874c0 queue: ictx=0x55c1b38754c0, req=0x7f596c008120
> 2019-07-26T10:06:30.125-0400 7f59bc942700 20 librbd::io::ImageRequestWQ: 0x55c1b38874c0 process: ictx=0x55c1b38754c0, req=0x7f596c008120
> 2019-07-26T10:06:30.125-0400 7f59bc942700 20 librbd::io::ImageRequest: 0x7f59bc93e570 send: aio_write: ictx=0x55c1b38754c0, completion=0x55c1b38d82d0
> 2019-07-26T10:06:30.125-0400 7f59bc942700 20 librbd::io::AioCompletion: 0x55c1b38d82d0 set_request_count: pending=1
> 2019-07-26T10:06:30.125-0400 7f59bc942700 20 librbd::Journal: 0x7f59a4001cf0 append_io_events: event=AioWrite, offset=514141, length=512, flush=0, tid=158

MahatiC commented:
@dillaman Thanks much! I pushed a fix yesterday and no longer see the crash. I'll investigate the lockdep issue.

MahatiC commented:
retest this please

MahatiC commented:
@dillaman Pushed a new version to resolve lockdep issue. Could you please verify/review and run this PR through teuthology? Thanks.

dillaman commented:
http://pulpito.ceph.com/jdillaman-2019-08-01_19:52:47-rbd-wip-jd-testing-distro-basic-smithi/

MahatiC commented:
@dillaman I don't see any memory leaks and also don't see the overlapping IO issue. But there's an assert failure in test_rbd_mirror.sh workunit for test `TestImageDeleter.Delete_Image_With_ProtectedSnapshots`. I'm not quite sure if it has to do with changes in this PR. Could you take a look? Thanks.

dillaman commented:
@MahatiC FYI -- I am out-of-office for the next two weeks, so it will be a while before I can assist.

MahatiC commented:
> @MahatiC FYI -- I am out-of-office for the next two weeks, so it will be a while before I can assist.

Okay, thanks for the info.

MahatiC commented:
@dillaman Could you run this version through Teuthology? I fixed a couple of issues that resulted after the rebase.

dillaman commented:
Sure, I’ll try to build it and run it through tomorrow after my current set of test runs complete.

dillaman commented:
@MahatiC can you please rebase the PR?

MahatiC commented:
retest this please

MahatiC commented:
@dillaman rebased. build failures being addressed by #29862 . Will rebase once that's merged.

MahatiC commented:
@dillaman I found a couple of errors in the last run. Fixed a locking issue in rbd mirror that was causing undefined behavior. Could you please run this through Teuthology?

dillaman commented:
@MahatiC Hmm -- the test run looks worse:

http://pulpito.ceph.com/jdillaman-2019-09-03_21:31:23-rbd-wip-jd-testing-distro-basic-smithi/

MahatiC commented:
@dillaman oops, all of them look related to the recently pushed rbd mirror change. I was trying to correct the explicit mutex unlock as `TestInstances.NotifyRemove` test was failing (reproducible when this test is run in repeat mode) due to a lock issue - https://github.com/ceph/ceph/blob/master/src/tools/rbd_mirror/Instances.cc#L133-L149

Could it just be that the test needs to change?

MahatiC commented:
@dillaman I modified the test and reverted the rbd mirror related change (only modified `lock_guard`  with `unique_lock` since there's a need to unlock/lock mutex). The test now doesn't fail in repeat mode. But I can't verify other impacts if any, locally. Please give this a run.

MahatiC commented:
@dillaman ping. Just a reminder.

dillaman commented on src/librbd/io/ImageDispatchSpec.cc:
```diff
@@ -131,6 +131,20 @@ uint64_t ImageDispatchSpec<I>::extents_length() {
+std::pair<uint64_t, uint64_t> ImageDispatchSpec<I>::get_image_extents() {
+   std::pair<uint64_t, uint64_t> image_extents = {0, 0};
+   if(this->m_image_extents.size()) {
+     image_extents = this->m_image_extents.front();
```
Nit: just return the `front()` here ... and avoid pre-initializing the worst-case above.

MahatiC replied:
Done

dillaman replied:
Yes, understood. Just return `front()` directly (since you have already verified it has at least a single entry). The flush case can directly return the `{0, 0}` pair.

MahatiC replied:
> The flush case can directly return the `{0, 0}` pair.

This is accessed in `process()` which means I would need a mechanism to identify the flush operation. Could you explain why such an initialization is bad, particularly here? I suppose it's fine to initialize elsewhere where this function is called?

dillaman replied:
The flush case doesn't matter, you are really just saying that it has no IO extents. The fact that you are pre-initializing a `{0, 0}` pair for all cases and then only using the result for flushes is somewhat arbitrary. Passing back the full image extents would avoid this since then block/unblock would see that there are no associated extents.

MahatiC replied:
Done. It now returns full extents vector.

dillaman commented on src/librbd/io/ImageDispatchSpec.h:
```diff
@@ -160,10 +164,10 @@ class ImageDispatchSpec {
   ImageDispatchSpec(ImageCtxT& image_ctx, AioCompletion* aio_comp,
                      Extents&& image_extents, Request&& request,
-                     int op_flags, const ZTracer::Trace& parent_trace)
+                     int op_flags, const ZTracer::Trace& parent_trace, uint64_t tid=0)
```
Nit: don't default parameters

MahatiC replied:
Done

dillaman commented on src/librbd/io/ImageRequestWQ.h:
```diff
@@ -107,6 +108,46 @@ class ImageRequestWQ
+  };
+
+  vector<BlockedIO> m_blocked_ios;
+  ceph::mutex ordering_lock;
```
Nit: all member variables shall start w/ `m_`

MahatiC replied:
Done

dillaman commented on src/librbd/io/ImageRequestWQ.h:
```diff
@@ -107,6 +108,46 @@ class ImageRequestWQ
+
+  ImageExtents m_in_flight_extents;
+
+  struct BlockedIO {
```
Nit: can we not just re-use `ImageDispatcSpec`?

MahatiC replied:
Done

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -311,6 +313,26 @@ void ImageRequestWQ<I>::aio_write(AioCompletion *c, uint64_t off, uint64_t len,
+    {
+      std::lock_guard locker{ordering_lock};
+      tid = ++m_last_tid;
+      ldout(cct, 20) << "queueing tid: " << tid << dendl;
```
Nit: why can't this logic get moved down into `start_in_flight_io`?

MahatiC replied:
Added a new function `handle_overlapping_io` to handle this, since `start_in_flight_io` in it's current form is needed while unblocking IOs .

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -311,6 +313,26 @@ void ImageRequestWQ<I>::aio_write(AioCompletion *c, uint64_t off, uint64_t len,
+  if (native_async && (m_image_ctx.non_blocking_aio || writes_blocked())) {
+    {
+      std::lock_guard locker{ordering_lock};
+      tid = ++m_last_tid;
```
Can `m_last_tid` just be an `std::atomic`?

MahatiC replied:
Done.

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -311,6 +313,26 @@ void ImageRequestWQ<I>::aio_write(AioCompletion *c, uint64_t off, uint64_t len,
+  uint64_t tid = 0;
+  if (native_async && (m_image_ctx.non_blocking_aio || writes_blocked())) {
+    {
+      std::lock_guard locker{ordering_lock};
```
Nit: why does this need a separate lock?

MahatiC replied:
The lock is for `m_queued_or_blocked_io_tids` and `m_blocked_ios` that are being used below. `m_last_tid` is now atomic

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -326,8 +348,12 @@ void ImageRequestWQ<I>::aio_write(AioCompletion *c, uint64_t off, uint64_t len,
   if (m_image_ctx.non_blocking_aio || writes_blocked()) {
+    {
+      std::lock_guard locker{ordering_lock};
+      m_queued_or_blocked_io_tids.insert(tid);
```
Nit: can this be moved to `queue`?

MahatiC replied:
Need those tids ordered in `unblock_flushes`

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -386,6 +412,19 @@ void ImageRequestWQ<I>::aio_flush(AioCompletion *c, bool native_async) {
     trace.event("init");
   }
 
+  if (native_async &&
```
Nit: same comment here

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -492,6 +531,110 @@ void ImageRequestWQ<I>::aio_compare_and_write(AioCompletion *c,
+    auto blocked_flush = *it;
+    ldout(cct, 20) << "unblocking flush: tid" << blocked_flush.first << dendl;
+    blocked_flush.second.c->init_time(util::get_image_ctx(&m_image_ctx), AIO_TYPE_FLUSH);
+    if (blocked_flush.second.native_async && m_image_ctx.event_socket.is_valid()) {
```
Nit: why couldn't this be set up in `aio_flush`?

MahatiC replied:
Done

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -492,6 +531,110 @@ void ImageRequestWQ<I>::aio_compare_and_write(AioCompletion *c,
+      queue(ImageDispatchSpec<I>::create_flush_request(
+          m_image_ctx, blocked_flush.second.c, FLUSH_SOURCE_USER, trace));
+    }
+    ordering_lock.lock();
```
Nit: odd to lock just to unlock in the next line

MahatiC replied:
It is being unlocked after the `while` loop. And lock is being acquired before processing more within in the loop.

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -492,6 +531,110 @@ void ImageRequestWQ<I>::aio_compare_and_write(AioCompletion *c,
+            m_image_ctx, blocked_io.c, {{blocked_io.off, blocked_io.len}}, bufferlist{blocked_io.bl},
+            blocked_io.op_flags, trace, blocked_io.tid));
+      }
+      ordering_lock.lock();
```
Nit: same comment here

MahatiC commented:
@dillaman Ack. Working on other write_ops and review comments.

MahatiC commented:
retest this please

MahatiC commented:
@dillaman Addressed comments and other write_ops. Please review.

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -492,6 +551,102 @@ void ImageRequestWQ<I>::aio_compare_and_write(AioCompletion *c,
+      }
+      ldout(cct, 20) << "unblocking off: " << extents.first << ", "
+                     << "len: " << extents.second << dendl;
+      ZTracer::Trace trace;
```
Nit: same comment here

MahatiC replied:
removed

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -492,6 +551,102 @@ void ImageRequestWQ<I>::aio_compare_and_write(AioCompletion *c,
+    ldout(cct, 20) << "unblocking flush: tid" << blocked_flush.first << dendl;
+
+    AioCompletion *aio_comp = blocked_flush.second->get_aio_comp();
+    if (!start_in_flight_io(aio_comp)) {
```
Nit: this code looks very similar to the unblock IO code below. Can they share a single helper method to reduce code duplication?

MahatiC replied:
Done

dillaman commented on src/librbd/io/ImageRequestWQ.h:
```diff
@@ -107,6 +108,25 @@ class ImageRequestWQ
+    ImageExtents* in_flight_image_extents, uint64_t object_off,
+    uint64_t object_len);
+
+  void unblock_overlapping_io(uint64_t off, uint64_t len, uint64_t tid);
```
Nit: pass `ImageDispatchSpec` instead of individual params

MahatiC replied:
This doesn't seem possible since after `req->send()` in `process()`, the image extents are not preserved. I'll probably have to hold onto a copy of `ImageDispatchSpec` before `req->send()` in order to achieve this. I think that would be ok?

dillaman replied:
Ack. I guess for this particular method it would make sense to pass the extents (just add a comment to explain that the extents are invalidated after the request is sent so that the purpose is clear).

MahatiC replied:
Done. Added comment.

dillaman commented on src/librbd/io/ImageRequestWQ.h:
```diff
@@ -107,6 +108,25 @@ class ImageRequestWQ
+  vector<ImageDispatchSpec<ImageCtxT>*> m_blocked_ios;
+  ceph::mutex m_ordering_lock;
+
+  bool block_overlapping_io(
```
Nit: pass `ImageDispatchSpec` instead of individual params

MahatiC replied:
It's probably now not possible with the modified code, since this check is now being made before creating a request

dillaman commented on src/librbd/io/ImageRequestWQ.h:
```diff
@@ -126,10 +146,13 @@ class ImageRequestWQ
 
-  void finish_queued_io(ImageDispatchSpec<ImageCtxT> *req);
+  void finish_queued_io(ImageDispatchSpec<ImageCtxT> *req, bool write_op);
+  void remove_in_flight_write_ios(uint64_t off, uint64_t len, bool write_op, uint64_t tid);
```
Nit: pass `ImageDispatchSpec` instead of individual params

MahatiC replied:
Done

dillaman commented on src/librbd/io/ImageRequestWQ.h:
```diff
@@ -126,10 +146,13 @@ class ImageRequestWQ
   void finish_in_flight_write();
 
   int start_in_flight_io(AioCompletion *c);
+  int handle_overlapping_io(uint64_t off, uint64_t len,
```
Nit: no need for off/len/tid since you are already passing the `ImageDispatchSpec`

MahatiC replied:
Done

MahatiC commented on src/librbd/io/ImageRequestWQ.h:
```diff
@@ -107,6 +108,25 @@ class ImageRequestWQ
+    uint64_t object_len);
+
+  void unblock_overlapping_io(uint64_t off, uint64_t len, uint64_t tid);
+  std::atomic<unsigned> m_last_tid { 0 };
```
Done

MahatiC commented on src/librbd/io/ImageRequestWQ.h:
```diff
@@ -107,6 +108,25 @@ class ImageRequestWQ
   std::atomic<unsigned> m_io_blockers { 0 };
   std::atomic<unsigned> m_io_throttled { 0 };
 
+  typedef interval_set<uint64_t> ImageExtents;
```
Done

MahatiC commented on src/librbd/io/ImageRequestWQ.h:
```diff
@@ -107,6 +108,25 @@ class ImageRequestWQ
+
+  ImageExtents m_in_flight_extents;
+
+  vector<ImageDispatchSpec<ImageCtxT>*> m_blocked_ios;
```
Done

MahatiC commented on src/librbd/io/ImageDispatchSpec.h:
```diff
@@ -146,6 +146,14 @@ class ImageDispatchSpec {
 
+  std::pair<uint64_t, uint64_t> get_image_extents();
+
+  AioCompletion* get_aio_comp() {
```
Done

MahatiC replied:
Done

MahatiC commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -94,7 +94,9 @@ ImageRequestWQ<I>::ImageRequestWQ(I *image_ctx, const string &name,
     m_lock(ceph::make_shared_mutex(
-      util::unique_lock_name("ImageRequestWQ<I>::m_lock", this))) {
+      util::unique_lock_name("ImageRequestWQ<I>::m_lock", this))),
+    m_ordering_lock(ceph::make_mutex(
```
Replaced with `m_lock`

MahatiC commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -320,14 +322,24 @@ void ImageRequestWQ<I>::aio_write(AioCompletion *c, uint64_t off, uint64_t len,
+  if (m_image_ctx.non_blocking_aio || writes_blocked()) {
+    req = ImageDispatchSpec<I>::create_write_request(
+              m_image_ctx, c, {{off, len}}, std::move(bl), op_flags, trace, tid);
+    if(native_async && !handle_overlapping_io(off, len, tid, req)) {
```
Done

MahatiC commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -398,10 +420,27 @@ void ImageRequestWQ<I>::aio_flush(AioCompletion *c, bool native_async) {
+      {
+        std::lock_guard locker{m_ordering_lock};
+        if(!m_queued_or_blocked_io_tids.empty()) {
+          ldout(cct, 20) << "queueing Flush, tid: " << tid << dendl;
```
Done

MahatiC commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -826,12 +1011,32 @@ void ImageRequestWQ<I>::finish_in_flight_write() {
 }
 
+template <typename I>
+int ImageRequestWQ<I>::handle_overlapping_io(uint64_t off, uint64_t len,
```
Done

MahatiC commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -492,6 +551,102 @@ void ImageRequestWQ<I>::aio_compare_and_write(AioCompletion *c,
+    return false;
+  }
+
+  if (!in_flight_image_extents->empty() && in_flight_image_extents->intersects(off, len)) {
```
I'm not quite sure I understand this, I moved the extents update into the previous loop. Was that the suggestion

MahatiC commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -492,6 +551,102 @@ void ImageRequestWQ<I>::aio_compare_and_write(AioCompletion *c,
+void ImageRequestWQ<I>::unblock_flushes(uint64_t tid) {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 20) << "ictx=" << &m_image_ctx << dendl;
+  ZTracer::Trace trace;
```
Done

MahatiC commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -492,6 +551,102 @@ void ImageRequestWQ<I>::aio_compare_and_write(AioCompletion *c,
+}
+
+template <typename I>
+void ImageRequestWQ<I>::unblock_overlapping_io(uint64_t off, uint64_t len, uint64_t tid) {
```
Done

MahatiC commented on src/test/librbd/io/test_mock_ImageRequestWQ.cc:
```diff
@@ -44,7 +44,7 @@ struct ImageDispatchSpec<librbd::MockTestImageCtx> {
       librbd::MockTestImageCtx &image_ctx, AioCompletion *aio_comp,
       Extents &&image_extents, bufferlist &&bl, int op_flags,
-      const ZTracer::Trace &parent_trace) {
+      const ZTracer::Trace &parent_trace, uint64_t tid=0) {
```
Done

MahatiC commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -398,10 +420,27 @@ void ImageRequestWQ<I>::aio_flush(AioCompletion *c, bool native_async) {
   if (m_image_ctx.non_blocking_aio || writes_blocked() || !writes_empty()) {
-    queue(ImageDispatchSpec<I>::create_flush_request(
-            m_image_ctx, c, FLUSH_SOURCE_USER, trace));
+    queue(req);
```
It now creates a new request in both queued cases. But in order to not have two of those requests at the same time, I had to make the `block_overlapping_io` check as a precondition.

MahatiC commented:
@dillaman Updated. Please review.

MahatiC commented:
@dillaman ping

dillaman commented on src/librbd/io/ImageDispatchSpec.h:
```diff
@@ -146,6 +146,14 @@ class ImageDispatchSpec {
     return (m_throttled_flag & RBD_QOS_MASK) == RBD_QOS_MASK;
   }
 
+  Extents get_image_extents();
```
Nit: 
```suggestion
  const Extents& get_image_extents() const;
```

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -320,14 +320,32 @@ void ImageRequestWQ<I>::aio_write(AioCompletion *c, uint64_t off, uint64_t len,
 
+  auto tid = ++m_last_tid;
+
+  if (m_image_ctx.non_blocking_aio || writes_blocked()) {
```
This all still seems racy to me. I think you should just unconditionally execute `ImageDispatchSpec<I>::create_XYZ`, then queue it for processing (remove the current logic which attempts to execute it in the same thread context immediately). 

After creating the request but prior to queueing it, record the request tid in the `m_queued_or_blocked_io_tids` collection. It seems like then you could avoid the need to `prepare_overlapping_io` since you could move that logic to the process path and only block IOs if overlapping IOs are actively being dispatched instead of blocking on queued IOs as well.

MahatiC replied:
hmm, if `ImageDispatchSpec<I>::create_XYZ` is executed unconditionally and down below by the time it reaches to `m_image_ctx.non_blocking_aio` and if it's false, wouldn't that lead to a memory leak?

With the conditional check now, it only ever creates a request where it can't be queued immediately (i.e. when it needs to be blocked). In other words, if this is executed:

```
if (blocked) {
  req = ImageDispatchSpec<I>::create_write_request(
              m_image_ctx, c, {{off, len}}, std::move(bl), op_flags, trace, tid);
}
```
then it returns (after `prepare_overlapping_io`) & won't reach the below lines:

```
 std::shared_lock owner_locker{m_image_ctx.owner_lock};
 if (m_image_ctx.non_blocking_aio || writes_blocked()) {
    queue(ImageDispatchSpec<I>::create_write_request(
```

MahatiC replied:
Additionally, if `ImageDispatchSpec<I>::create_XYZ` is executed w/o a condition, and if the request is a blocking IO (`non_blocking_aio=false`), then it seems `ImageRequest<I>::aio_write` can't access the bufferlist due to `std::move` semantics

dillaman replied:
The bug right now is that you test `m_image_ctx.non_blocking_aio || writes_blocked` under two different `if` conditions so the results might change. This will either leave `req` uninitialized when it attempts to queue it or it will leak memory if it directly executes it. 

In the cause where you would normally directly invoke it, just directly process the pre-created request.

MahatiC replied:
> This will either leave req uninitialized when it attempts to queue

In the current code, `req` variable is not used in `queue`. It is executed directly. However I do see that, if the first if  condition `m_image_ctx.non_blocking_aio` is false and the second if condition is true, code won't be checking for overlapping IOs for that request.

> In the cause where you would normally directly invoke it, just directly process the pre-created request.

I will still need to execute `ImageDispatchSpec<I>::create_XYZ` under `m_image_ctx.non_blocking_aio || writes_blocked` condition. Perhaps something like below?

```
if (m_image_ctx.non_blocking_aio || writes_blocked()) {
  ImageDispatchSpec<I> *req = ImageDispatchSpec<I>::create_write_request(
              m_image_ctx, c, {{off, len}}, std::move(bl), op_flags, trace, tid);

  if (!prepare_overlapping_io(blocked, req)) {
      return;
  }
  std::shared_lock owner_locker{m_image_ctx.owner_lock};
  queue(req);
```

dillaman replied:
Wouldn't that just break the `else` path where it's directly executed? i.e. it might execute when it should have been blocked?

MahatiC replied:
oh you meant something like below?

```
ImageDispatchSpec<I> *req = ImageDispatchSpec<I>::create_write_request(
              m_image_ctx, c, {{off, len}}, std::move(bl), op_flags, trace, tid);

// add logic for tids into `m_queued_or_blocked_io_tids` and handle in `process`

std::shared_lock owner_locker{m_image_ctx.owner_lock};
if (m_image_ctx.non_blocking_aio || writes_blocked()) {
  queue(req);
else {
  ...
  req->send(); 
  ....
}
```

dillaman replied:
Yup -- but in a perfect world you would be able to call `process` instead of `req->send()` (or both `process` and this could call the same helper

MahatiC replied:
Done

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -826,12 +1040,33 @@ void ImageRequestWQ<I>::finish_in_flight_write() {
+
+   m_queued_or_blocked_io_tids.insert(m_last_tid.load());
+   if (blocked) {
+     Extents extents = req->get_image_extents();
```
```suggestion
     auto& extents = req->get_image_extents();
```

MahatiC replied:
Done

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -879,7 +1114,11 @@ void ImageRequestWQ<I>::fail_in_flight_io(
-  finish_queued_io(req);
+  bool write_op = req->is_write_op();
+  uint64_t tid = req->get_tid();
+  Extents extents = req->get_image_extents();
```
```suggestion
  auto& extents = req->get_image_extents();
```

MahatiC replied:
Done

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -320,18 +320,26 @@ void ImageRequestWQ<I>::aio_write(AioCompletion *c, uint64_t off, uint64_t len,
   }
 
-  if (!start_in_flight_io(c)) {
+ if (!start_in_flight_io(c)) {
```
```suggestion
  if (!start_in_flight_io(c)) {
```

MahatiC replied:
Done

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -320,18 +320,26 @@ void ImageRequestWQ<I>::aio_write(AioCompletion *c, uint64_t off, uint64_t len,
+  auto tid = ++m_last_tid;
+
+  {
+   std::lock_guard locker{m_lock};
```
Nit: indentation seems off

MahatiC replied:
Done

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -363,14 +371,21 @@ void ImageRequestWQ<I>::aio_discard(AioCompletion *c, uint64_t off,
 
+  auto tid = ++m_last_tid;
+  {
+   std::lock_guard locker{m_lock};
```
Nit: indentation seems off

MahatiC replied:
Done

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -436,14 +465,22 @@ void ImageRequestWQ<I>::aio_writesame(AioCompletion *c, uint64_t off,
+  auto tid = ++m_last_tid;
+
+  {
+   std::lock_guard locker{m_lock};
```
Nit: indentation seems off

MahatiC replied:
Done

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -477,21 +514,116 @@ void ImageRequestWQ<I>::aio_compare_and_write(AioCompletion *c,
+  auto tid = ++m_last_tid;
+
+  {
+   std::lock_guard locker{m_lock};
```
Nit: indentation seems off

MahatiC replied:
Done

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -787,26 +919,90 @@ void *ImageRequestWQ<I>::_void_dequeue() {
+  process_io(req, true);
+
+  finish_queued_io(write_op);
   delete req;
```
Nit: move this to `process_io` or else you will have a memory leak on the blocking IO path

MahatiC replied:
Done

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -787,26 +919,90 @@ void *ImageRequestWQ<I>::_void_dequeue() {
+  {
+    std::lock_guard locker{m_lock};
+    if (write_op) {
+      if (length) {
```
```suggestion
      if (length > 0) {
```

MahatiC replied:
Done

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -477,21 +514,116 @@ void ImageRequestWQ<I>::aio_compare_and_write(AioCompletion *c,
+  ldout(cct, 20) << "ictx=" << &m_image_ctx
+                 << "off: " << off << " len: " << len <<dendl;
+
+  if(!len) {
```
```suggestion
  if(len == 0) {
```

MahatiC replied:
Done

dillaman commented:
jenkins test submodules

xuqing-er commented:
@dillaman @MahatiC I would like to ask why it is necessary to wait for the previous write operation to complete before continuing to write, but this part is missing in the pacific version
